### ISSUE TITLE
ErrorBoundary unable to handle Errors throw in Async function

### ISSUE THREAD

**Issue by seetdev** [kentcdodds/react-suspense#43]
This might be related to #37 but the following line of code will not be handled by the ErrorBoundary
https://github.com/kentcdodds/react-suspense/blob/c6a440d3b04b5ba6693423b6c8e65f5fd9c8dbaa/src/pokemon.js#L77

And I believe the following will not work as well although I was not able to trigger it.
https://github.com/kentcdodds/react-suspense/blob/c6a440d3b04b5ba6693423b6c8e65f5fd9c8dbaa/src/pokemon.js#L68-L70

The reason is ErrorBoundaries cannot catch errors thrown in Asynchronous code and since those lines live in the then block it does not handle it well.
```
Note

Error boundaries do not catch errors for:
- Event handlers (learn more)
- Asynchronous code (e.g. setTimeout or requestAnimationFrame callbacks)
- Server side rendering
- Errors thrown in the error boundary itself (rather than its children)
```

The fix that I found was to it return an object with a message property. 
```
        return Promise.reject({
           message: `No pokemon with the name "${name}"`
        });
```
It needs to be an object with the message property since ErrorBoundary expects to get that at least for this implementation.
https://github.com/kentcdodds/react-suspense/blob/c6a440d3b04b5ba6693423b6c8e65f5fd9c8dbaa/src/pokemon.js#L272-L282

It might not work if other properties from the error argument is access by the code handling the Error Boundary.  

I have a codesandbox that displays the behavior and also the fix if you replace the return statement by uncommenting.
[Codesandbox](https://codesandbox.io/s/hardcore-http-9sc4w)

The interesting thing is this error does not show up in when you run `01.js` in either `final` or `exercise`. I believe it is due to `02.js` putting the resource using the `useState` hook.

I will be happy to add in a PR if the proposed change is the correct way to do this.

PS. Dan has a way to handle those as an abstraction and I might try it out next time.

https://github.com/facebook/react/issues/14981#issuecomment-468460187

**Comment by intrueder** [comment]
I checked it, and it works as intended (after dismissing an error overlay from React in development mode)

![image](https://user-images.githubusercontent.com/182339/101936409-348ce600-3be9-11eb-8cd5-d0af596a34f4.png)

**Comment by seetdev** [comment]
Oh ok. Now I see what is going on now. It is due to it being in development mode. When running this on local, there was a flash that shows the ErrorBoundary rendering and then the following screenshot. Funny thing was in `01.js` the same behavior does not show up. You always learn something new with EpicReact

<img width="455" alt="Screen Shot 2020-12-11 at 1 21 11 PM" src="https://user-images.githubusercontent.com/35116035/101940031-db568f80-3bb3-11eb-9ddb-52754c5ff6e1.png">

**Comment by intrueder** [comment]
You are welcome!
I agree, overlay design is confusing, and we are not alone. Here is an issue that explains it a bit: https://github.com/facebook/create-react-app/issues/3627

**Comment by kentcdodds** [comment]
I chimed in with my thoughts on an issue and Dan responded about it here: https://github.com/facebook/create-react-app/issues/6530#issuecomment-662914362

**Comment by rpedroni-sf** [comment]
Been fighting with this for quite some workshops now and finally found this answer, thanks!
Might be a good thing to mention in an updated version - that is, if you haven't mentioned it yet and I just missed it :P

**Comment by kentcdodds** [comment]
@rpedroni-sf do you have a suggestion of where we could mention this so people don't miss it?

**Comment by rpedroni-sf** [comment]
@kentcdodds Maybe expand "Concurrent Mode" with an extra, something on the lines of "Extra - Warning about Errors"? 
Think it'd get people's attention enough since it's a really short one

**Comment by seetdev** [comment]
It has been sometime since I completed the lessons and I tried finding a lesson on ErrorBoundaries but could not locate it.
Just a suggestion if it was not in the workshops. Maybe a lesson on ErrorBoundaries by itself with this note might be a good resource. This especially if ErrorBoundaries are suppose to be part of React anyway.


I am not sure about @rpedroni-sf  but I believed I was caught by this due to being familiar with how errors get displayed by IIS in an error involving a .NET web application. Not sure about it but maybe the overlay was created by someone who did .NET in his previous incarnation.

![image](https://user-images.githubusercontent.com/35116035/105106736-6fa1f500-5a84-11eb-934b-4c91a3df466b.png)
